--- ../src-base/minecraft/net/minecraft/entity/EnumCreatureType.java
+++ ../src-work/minecraft/net/minecraft/entity/EnumCreatureType.java
@@ -9,6 +9,6 @@
 public enum EnumCreatureType
 {
-    monster,
-    creature,
-    ambient,
-    waterCreature;
+    monster(IMob.class, 70, Material.air, false, false),
+    creature(EntityAnimal.class, 10, Material.air, true, true),
+    ambient(EntityAmbientCreature.class, 15, Material.air, true, false),
+    waterCreature(EntityWaterMob.class, 5, Material.water, true, false);
@@ -55,10 +55,2 @@
     }
-
-    static
-    {
-        monster = new EnumCreatureType("monster", 0, IMob.class, 70, Material.field_151579_a, false, false);
-        creature = new EnumCreatureType("creature", 1, EntityAnimal.class, 10, Material.field_151579_a, true, true);
-        ambient = new EnumCreatureType("ambient", 2, EntityAmbientCreature.class, 15, Material.field_151579_a, true, false);
-        waterCreature = new EnumCreatureType("waterCreature", 3, EntityWaterMob.class, 5, Material.field_151586_h, true, false);
-    }
 }
