--- ../src-base/minecraft/net/minecraft/world/gen/structure/StructureVillagePieces.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/StructureVillagePieces.java
@@ -4,5 +4,7 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Random;
+import cpw.mods.fml.common.eventhandler.Event.Result;
+import cpw.mods.fml.common.registry.VillagerRegistry;
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
@@ -16,6 +18,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.biome.BiomeGenBase;
 import net.minecraft.world.biome.WorldChunkManager;
+import net.minecraftforge.common.ChestGenHooks;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.terraingen.BiomeEvent;
+import static net.minecraftforge.common.ChestGenHooks.*;
 
 public class StructureVillagePieces
 {
@@ -50,6 +56,8 @@
         arraylist.add(new StructureVillagePieces.PieceWeight(StructureVillagePieces.Field2.class, 3, MathHelper.func_76136_a(p_75084_0_, 2 + p_75084_1_, 4 + p_75084_1_ * 2)));
         arraylist.add(new StructureVillagePieces.PieceWeight(StructureVillagePieces.House2.class, 15, MathHelper.func_76136_a(p_75084_0_, 0, 1 + p_75084_1_)));
         arraylist.add(new StructureVillagePieces.PieceWeight(StructureVillagePieces.House3.class, 8, MathHelper.func_76136_a(p_75084_0_, 0 + p_75084_1_, 3 + p_75084_1_ * 2)));
+        VillagerRegistry.addExtraVillageComponents(arraylist, p_75084_0_, p_75084_1_);
+
         Iterator iterator = arraylist.iterator();
 
         while (iterator.hasNext())
@@ -123,6 +131,10 @@
         {
             object = StructureVillagePieces.House3.func_74921_a(p_75083_0_, p_75083_2_, p_75083_3_, p_75083_4_, p_75083_5_, p_75083_6_, p_75083_7_, p_75083_8_);
         }
+        else
+        {
+            object = VillagerRegistry.getVillageComponent(p_75083_1_, p_75083_0_ , p_75083_2_, p_75083_3_, p_75083_4_, p_75083_5_, p_75083_6_, p_75083_7_, p_75083_8_);
+        }
 
         return (StructureVillagePieces.Village)object;
     }
@@ -925,7 +937,7 @@
                     if (p_74875_3_.func_78890_b(j, i, k))
                     {
                         this.field_74917_c = true;
-                        this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 5, 1, 5, field_74918_a, 3 + p_74875_2_.nextInt(6));
+                        this.func_74879_a(p_74875_1_, p_74875_3_, p_74875_2_, 5, 1, 5, ChestGenHooks.getItems(VILLAGE_BLACKSMITH, p_74875_2_), ChestGenHooks.getCount(VILLAGE_BLACKSMITH, p_74875_2_));
                     }
                 }
 
@@ -1422,6 +1434,7 @@
             public List field_74932_i = new ArrayList();
             public List field_74930_j = new ArrayList();
             private static final String __OBFID = "CL_00000527";
+            public BiomeGenBase biome;
 
             public Start()
             {
@@ -1433,6 +1446,7 @@
                 this.field_74928_c = p_i2104_7_;
                 BiomeGenBase biomegenbase = p_i2104_1_.func_76935_a(p_i2104_4_, p_i2104_5_);
                 this.field_74927_b = biomegenbase == BiomeGenBase.field_76769_d || biomegenbase == BiomeGenBase.field_76786_s;
+                this.biome = biomegenbase;
             }
 
             public WorldChunkManager func_74925_d()
@@ -1493,6 +1507,7 @@
             private int field_74896_a;
             private boolean field_143014_b;
             private static final String __OBFID = "CL_00000531";
+            private StructureVillagePieces.Start startPiece;
 
             public Village()
             {
@@ -1503,6 +1518,7 @@
                 if (p_i2107_1_ != null)
                 {
                     this.field_143014_b = p_i2107_1_.field_74927_b;
+                    startPiece = p_i2107_1_;
                 }
             }
 
@@ -1616,6 +1632,9 @@
 
             protected Block func_151558_b(Block p_151558_1_, int p_151558_2_)
             {
+                BiomeEvent.GetVillageBlockID event = new BiomeEvent.GetVillageBlockID(startPiece == null ? null : startPiece.biome, p_151558_1_, p_151558_2_);
+                MinecraftForge.TERRAIN_GEN_BUS.post(event);
+                if (event.getResult() == Result.DENY) return event.replacement;
                 if (this.field_143014_b)
                 {
                     if (p_151558_1_ == Blocks.field_150364_r || p_151558_1_ == Blocks.field_150363_s)
@@ -1654,6 +1673,9 @@
 
             protected int func_151557_c(Block p_151557_1_, int p_151557_2_)
             {
+                BiomeEvent.GetVillageBlockMeta event = new BiomeEvent.GetVillageBlockMeta(startPiece == null ? null : startPiece.biome, p_151557_1_, p_151557_2_);
+                MinecraftForge.TERRAIN_GEN_BUS.post(event);
+                if (event.getResult() == Result.DENY) return event.replacement;
                 if (this.field_143014_b)
                 {
                     if (p_151557_1_ == Blocks.field_150364_r || p_151557_1_ == Blocks.field_150363_s)
