--- ../src-base/minecraft/net/minecraft/network/NetworkManager.java
+++ ../src-work/minecraft/net/minecraft/network/NetworkManager.java
@@ -282,1 +282,3 @@
+               if(net.minecraftforkage.internal.FakeLatencyHandler.isEnabled())
+                   p_initChannel_1_.pipeline().addFirst("fake_latency", new net.minecraftforkage.internal.FakeLatencyHandler());
                 p_initChannel_1_.pipeline().addLast("packet_handler", networkmanager);
