--- ../src-base/minecraft/net/minecraft/client/shader/ShaderLoader.java
+++ ../src-work/minecraft/net/minecraft/client/shader/ShaderLoader.java
@@ -87,6 +87,6 @@
     @SideOnly(Side.CLIENT)
     public static enum ShaderType
     {
-        VERTEX,
-        FRAGMENT;
+        VERTEX("vertex", ".vsh", OpenGlHelper.field_153209_q),
+        FRAGMENT("fragment", ".fsh", OpenGlHelper.field_153210_r);
 
@@ -125,8 +125,2 @@
-
-        static
-        {
-            VERTEX = new ShaderLoader.ShaderType("VERTEX", 0, "vertex", ".vsh", OpenGlHelper.field_153209_q);
-            FRAGMENT = new ShaderLoader.ShaderType("FRAGMENT", 1, "fragment", ".fsh", OpenGlHelper.field_153210_r);
-        }
     }
 }
