--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
@@ -219,8 +219,7 @@
 
                             if (this.field_145957_n[1].field_77994_a == 0)
                             {
-                                Item item = this.field_145957_n[1].func_77973_b().func_77668_q();
-                                this.field_145957_n[1] = item != null ? new ItemStack(item) : null;
+                                this.field_145957_n[1] = field_145957_n[1].func_77973_b().getContainerItem(field_145957_n[1]);
                             }
                         }
                     }
@@ -265,7 +264,11 @@
         else
         {
             ItemStack itemstack = FurnaceRecipes.func_77602_a().func_151395_a(this.field_145957_n[0]);
-            return itemstack == null ? false : (this.field_145957_n[2] == null ? true : (!this.field_145957_n[2].func_77969_a(itemstack) ? false : (this.field_145957_n[2].field_77994_a < this.func_70297_j_() && this.field_145957_n[2].field_77994_a < this.field_145957_n[2].func_77976_d() ? true : this.field_145957_n[2].field_77994_a < itemstack.func_77976_d())));
+            if (itemstack == null) return false;
+            if (this.field_145957_n[2] == null) return true;
+            if (!this.field_145957_n[2].func_77969_a(itemstack)) return false;
+            int result = field_145957_n[2].field_77994_a + itemstack.field_77994_a;
+            return result <= func_70297_j_() && result <= this.field_145957_n[2].func_77976_d(); //Forge BugFix: Make it respect stack sizes properly.
         }
     }
 
@@ -281,7 +284,7 @@
             }
             else if (this.field_145957_n[2].func_77973_b() == itemstack.func_77973_b())
             {
-                ++this.field_145957_n[2].field_77994_a;
+                this.field_145957_n[2].field_77994_a += itemstack.field_77994_a; // Forge BugFix: Results may have multiple items
             }
 
             --this.field_145957_n[0].field_77994_a;
@@ -301,6 +304,9 @@
         }
         else
         {
+        	int moddedBurnTime = net.minecraftforge.event.ForgeEventFactory.getFuelBurnTime(p_145952_0_);
+        	if (moddedBurnTime >= 0) return moddedBurnTime;
+        	
             Item item = p_145952_0_.func_77973_b();
 
             if (item instanceof ItemBlock && Block.func_149634_a(item) != Blocks.field_150350_a)
@@ -321,7 +323,15 @@
                 }
             }
 
-            return item instanceof ItemTool && ((ItemTool)item).func_77861_e().equals("WOOD") ? 200 : (item instanceof ItemSword && ((ItemSword)item).func_150932_j().equals("WOOD") ? 200 : (item instanceof ItemHoe && ((ItemHoe)item).func_77842_f().equals("WOOD") ? 200 : (item == Items.field_151055_y ? 100 : (item == Items.field_151044_h ? 1600 : (item == Items.field_151129_at ? 20000 : (item == Item.func_150898_a(Blocks.field_150345_g) ? 100 : (item == Items.field_151072_bj ? 2400 : 0)))))));
+            if (item instanceof ItemTool && ((ItemTool)item).func_77861_e().equals("WOOD")) return 200;
+            if (item instanceof ItemSword && ((ItemSword)item).func_150932_j().equals("WOOD")) return 200;
+            if (item instanceof ItemHoe && ((ItemHoe)item).func_77842_f().equals("WOOD")) return 200;
+            if (item == Items.field_151055_y) return 100;
+            if (item == Items.field_151044_h) return 1600;
+            if (item == Items.field_151129_at) return 20000;
+            if (item == Item.func_150898_a(Blocks.field_150345_g)) return 100;
+            if (item == Items.field_151072_bj) return 2400;
+            return cpw.mods.fml.common.registry.GameRegistry.getFuelValue(p_145952_0_);
         }
     }
 
